### PR TITLE
Firewall active zones 

### PR DESCRIPTION
**Pull request by Gundersanne**

### CODE REVIEW

File: pkg/networkmanager/firewall-client.js [hunk 1]
```diff
@@ -25,7 +25,10 @@ var firewall = {
     enabled: false,
     readonly: true,
     services: {},
-    enabledServices: new Set()
+    enabledServices: new Set(),
+    zones: {},
```
Comment (martinpitt) [inline pkg/networkmanager/firewall-client.js]: Can you please add a comment what the structure of this is?

File: pkg/networkmanager/firewall-client.js [hunk 2]
```diff
@@ -158,6 +180,38 @@ function fetchServiceInfos(services) {
     });
 }
 
+function fetchZoneInfos(zones) {
+    let zoneNames = Array.isArray(zones) ? zones : Object.keys(zones);
```
Comment (martinpitt) [inline pkg/networkmanager/firewall-client.js]: Does it realistically happen that this gets called on both arrays and lists? This is a bit confusing.

File: pkg/networkmanager/firewall.jsx [hunk 3]
```diff
@@ -94,6 +98,23 @@ function ServiceRow(props) {
                        tabRenderers={tabs} />;
 }
 
+function ZoneRow(props) {
+    let columns = [
+        { name: props.zone.name, header: true },
+        <div>
+            { props.zone.id === firewall.defaultZone ? <span className="fa fa-check">default</span> : '' }
+        </div>,
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: Do these really need to be `div`s for CSS purposes?

File: pkg/networkmanager/firewall-client.js [hunk 4]
```diff
@@ -50,6 +51,15 @@ function initFirewalldDbus() {
             return;
         }
 
+        /* As certain dbus signal callbacks might change the firewall frequently
+         * in a short period of time, prevent rapid succession of renders by
+         * debouncing the ('changed') event */
+        firewall.debouncedEvent = debounce(300, event => firewall.dispatchEvent(event));
```
Comment (martinpitt) [inline pkg/networkmanager/firewall-client.js]: The debouncing could also be split out into a separate commit/PR. (Not a strong opinion, just a random idea)

File: pkg/networkmanager/firewall-client.js [hunk 5]
```diff
@@ -277,36 +292,56 @@ firewall.createService = (service, name, ports) => {
...
                                'org.fedoraproject.FirewallD1.zone',
-                               'addService', ['', service, 0])
-            .then(reply => getDefaultZonePath())
-            .then(path => firewalld_dbus.call(path, 'org.fedoraproject.FirewallD1.config.zone',
+                               'addService', [zone, service, 0])
+            .then(reply => firewalld_dbus.call('/org/fedoraproject/FirewallD1/config',
+                                               'org.fedoraproject.FirewallD1.config',
+                                               'getZoneByName', [zone]))
+            .then(path => firewalld_dbus.call(path[0], 'org.fedoraproject.FirewallD1.config.zone',
                                               'addService', [service]))
```
Comment (martinpitt) [inline pkg/networkmanager/firewall-client.js]: Does that also cause the service to be added to the runtime config, or  just the persistent one? I think it should also do runtime, for consistency. Can you please explain the duality and strategy wrt. runtime vs. persistent in the commit message?

File: pkg/networkmanager/firewall.jsx [hunk 6]
```diff
@@ -181,6 +181,7 @@ class AddServicesModal extends React.Component {
             custom_udp_ports: [],
             custom_tcp_value: "",
             custom_udp_value: "",
+            zones: firewall.activeZones.size === 1 ? [firewall.defaultZone] : [],
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: Please add a comment that this is for a convenient default if there's just one zone.

File: pkg/networkmanager/firewall.jsx [hunk 7]
```diff
@@ -204,9 +206,9 @@ class AddServicesModal extends React.Component {
     save() {
         if (this.state.custom) {
             firewall.createService(this.state.custom_id, this.state.custom_name, this.createPorts())
-                    .then(firewall.enableService(this.state.custom_id));
+                    .then(firewall.enableService(this.state.zones.concat(), this.state.custom_id));
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: These functions shouldn't modify their arguments, so copying the list should not be necessary

File: pkg/networkmanager/firewall.jsx [hunk 8]
```diff
@@ -374,6 +376,15 @@ class AddServicesModal extends React.Component {
         });
     }
 
+    onToggleZone(event) {
+        let zone = event.target.value;
+        this.setState(state => {
+            if (state.zones.indexOf(zone) === -1)
+                return { zones: state.zones.concat(zone) };
+            return { zones: state.zones.filter(z => z !== zone) };
+        });
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: Might be easier to make `zones` a Set.

File: pkg/networkmanager/firewall.jsx [hunk 9]
```diff
@@ -418,6 +435,19 @@ class AddServicesModal extends React.Component {
                 </Modal.Header>
                 <div id="cockpit_modal_dialog">
                     <Modal.Body id="add-services-dialog">
+                        { firewall.activeZones.size > 1 &&
+                            <React.Fragment>
+                                <form className="ct-form-layout horizontal">
+                                    <label htmlFor="zone-input">{ _("Add services to following zones:") }</label>
+                                    <fieldset id="zone-input">
+                                        { Array.from(firewall.activeZones).map(z =>
```
Comment (martinpitt) [inline pkg/networkmanager/firewall.jsx]: This array should be sorted, both for better presentation and for more predictability in the tests.

File: test/verify/check-networking-firewall [hunk 10]
```diff
@@ -138,15 +152,22 @@ class TestFirewall(MachineCase):
         m = self.machine
 
         self.login_and_go("/network/firewall")
+        b.wait_in_text("#zones-listing", "Public")
+        b.wait_in_text("#zones-listing", "default")
 
         def removeService(service):
             b.click("tr[data-row-id='{}'] .btn.pficon-delete".format(service))
+            if b.is_present("#remove-services-dialog-body #zone-input input:not(:checked)"):
```
Comment (martinpitt) [inline test/verify/check-networking-firewall]: This kind of dynamic behaviour makes tests less predictable and understandable. Please rather turn this into an explicit argument if you want to use it for diifferent cases.

File: test/verify/check-networking-firewall [hunk 11]
```diff
@@ -138,15 +152,22 @@ class TestFirewall(MachineCase):
...
+                b.click("#remove-services-dialog-body #zone-input input:not(:checked)")
+            b.click("#remove-services-dialog button.btn-primary")
             b.wait_not_present("tr[data-row-id='{}']".format(service))
 
         # click on the "Add Services" button
         b.wait_present("caption button.btn-primary:enabled")
         b.click("caption button.btn-primary")
         b.wait_present("#cockpit_modal_dialog .list-view-pf input[data-id='pop3']")
+        if m.image not in ["rhel-8-0", "rhel-8-1"]:
```
Comment (martinpitt) [inline test/verify/check-networking-firewall]: This should come with a comment that explains what's special here. E. g. a bugzilla. Same for the other special cases below.
This also needs to be updated for rhel-8-0-distropkg.

File: test/verify/check-networking-firewall [hunk 12]
```diff
@@ -218,6 +248,8 @@ class TestFirewall(MachineCase):
             b.click("#add-services-dialog input[value='ports']")
             b.wait_present("#tcp-ports")
             b.wait_visible("#tcp-ports")
+            if b.is_present("#add-services-dialog input[value='public']"):
```
Comment (martinpitt) [inline test/verify/check-networking-firewall]: Dito for "unpredictable dynamic behaviour" above.

File: test/verify/check-networking-firewall [hunk 13]
```diff
@@ -280,7 +313,47 @@ class TestFirewall(MachineCase):
         set_field("#tcp-ports", "80-90-100", "")
         check_error("Invalid range")
 
+    @skipImage("Filtering got fixed, and IDs changed to data-id in 184", "rhel-7-6-distropkg")
```
Comment (martinpitt) [inline test/verify/check-networking-firewall]: needs to handle rhel-8-0-distropkg

File: test/verify/check-networking-firewall [hunk 14]
```diff
@@ -280,7 +313,47 @@ class TestFirewall(MachineCase):
...
+
+        self.login_and_go("/network/firewall")
+        b.wait_present("#zones-listing:contains({})".format(workSource))
+        b.wait_present("#zones-listing:contains(Work)")
+
+        addServiceToZone("pop3", "work")
+        b.wait_present("tr[data-row-id='pop3']:contains(Work)")
+        self.assertNotIn("Public", b.text("tr[data-row-id='pop3']"))
+        addServiceToZone("pop3", "public")
+        b.wait_present("tr[data-row-id='pop3']:contains(Public)")
```
Comment (martinpitt) [inline test/verify/check-networking-firewall]: Can you please add a `firewall-cmd` CLI check that the state is really as expected?

File: test/verify/check-networking-firewall [hunk 15]
```diff
@@ -280,7 +313,47 @@ class TestFirewall(MachineCase):
...
+        b = self.browser
+        m = self.machine
 
+        workSource = "192.168.1.0/24"
+        m.execute("firewall-cmd --permanent --zone=work --change-source={}".format(workSource))
+        m.execute("firewall-cmd --reload")
+
+        def addServiceToZone(service, zone):
+            b.click("caption button.btn-primary")
+            b.wait_present("#add-services-dialog")
```
Comment (martinpitt) [inline test/verify/check-networking-firewall]: Unnecessary, click() and other actions wait for their selector to be present now.

File: pkg/networkmanager/firewall-client.js (hunk 5 above)
Comment (Gundersanne) [reply pkg/networkmanager/firewall-client.js]: Both weirdly enough, if you add a custom service to the runtime configuration it'll cause an error.

I'll update the commit message.

File: pkg/networkmanager/firewall.jsx (hunk 8 above)
Comment (Gundersanne) [reply pkg/networkmanager/firewall.jsx]: I tried this but it turned out to be more verbose with copying it first to modify it etc... 

Plus this exact logic is used in another place of the firewall.

File: pkg/networkmanager/firewall-client.js (hunk 5 above)
Comment (martinpitt) [reply pkg/networkmanager/firewall-client.js]: So by that logic, why doesn't this get enabled in the runtime config here, and only in the persistent config?

Comment (Gundersanne) [reply pkg/networkmanager/firewall-client.js]: I must've implemented this badly when I tried it, indeed you can add it to the runtime config and then to the persistent one, although somehow it seems to be active right away regardless.

Anyway let's be consistent and add it to runtime, then permanent.

**Comment by martinpitt** [comment]
One remaining [test failure on rhel-8-1](https://209.132.184.41:8493/logs/pull-11745-20190507-152626-b47bda78-verify-rhel-8-1/log.html#89)

File: pkg/networkmanager/firewall-client.js (hunk 5 above)
Comment (martinpitt) [reply pkg/networkmanager/firewall-client.js]: That sounds like a bug -- it would mean that adding a service to a zone isn't immediately effective? If it is, then this  should be validated with a CLI test (`iptables -L`, or at least with `firewall-cmd`)

File: test/verify/check-networking-firewall (hunk 14 above)
Comment (Gundersanne) [reply test/verify/check-networking-firewall]: add this in `addServiceToZone` helper
